File: javalin-openapi-processor/src/main/java/net/dzikoysk/openapi/processor/annotations/OpenApiLoader.java
package net.dzikoysk.openapi.processor.annotations;

import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Set;

public final class OpenApiLoader {

    public static Collection<OpenApiInstance> loadAnnotations(Set<? extends TypeElement> annotations, Elements elements, Types types, RoundEnvironment roundEnv) {
        Collection<OpenApiInstance> openApiAnnotations = new ArrayList<>();

        for (TypeElement annotation : annotations) {
            Set<? extends Element> annotatedElements = roundEnv.getElementsAnnotatedWith(annotation);

            for (Element annotatedElement : annotatedElements) {
                AnnotationMirror openApiAnnotationMirror = annotatedElement.getAnnotationMirrors().get(0);
                OpenApiInstance openApiInstance = new OpenApiInstance(openApiAnnotationMirror);
                openApiAnnotations.add(openApiInstance);
                // types.asElement(openApiInstance.responses()[0].content()[0].from()).getEnclosedElements()
            }
        }

        return openApiAnnotations;
    }

}

File: javalin-openapi-processor/src/main/kotlin/net/dzikoysk/openapi/processor/annotations/OpenApiLoader.kt
package net.dzikoysk.openapi.processor.annotations

import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements
import javax.lang.model.util.Types

internal object OpenApiLoader {

    fun loadAnnotations(annotations: Set<TypeElement>, elements: Elements, types: Types, roundEnv: RoundEnvironment): Collection<OpenApiInstance> {
        val openApiAnnotations: MutableCollection<OpenApiInstance> = ArrayList()

        for (annotation in annotations) {
            val annotatedElements = roundEnv.getElementsAnnotatedWith(annotation)

            for (annotatedElement in annotatedElements) {
                val openApiAnnotationMirror = annotatedElement.annotationMirrors[0]
                val openApiInstance = OpenApiInstance(openApiAnnotationMirror)
                openApiAnnotations.add(openApiInstance)
            }
        }

        return openApiAnnotations
    }

}